[Kernel]Add streamK for block-quantized CUTLASS kernels

Hongbosherlock opened the pull request:
Thanks to https://github.com/vllm-project/vllm/pull/11868

Currently only supports scale_a block shapes of 1x128 and scale_b block shapes of 128x128 (for deepseek v3)

This CUTLASS GEMM is slower when K is very large, such as `m128-n1536-k7168` in deepseek v3.
To address this, I explored using both `Stream-K` and `Split-K `strategies to accelerate the GEMM computation.

Among these approaches, `Stream-K` demonstrated superior performance in most cases, achieving over **60% performance improvement** compared to the baseline.

To achieve overall better performance,  **use `Stream-K` only when `K > 3N`.**

Below are the test results based on the GEMM shape of DeepSeek-V3, tested on H800 with TP-4.

```
[----------------- scaled-torch.float8_e4m3fn-gemm ------------------]
                           |  cutlass_fp8_fp8_fp16_scaled_mm_blockwise
  -------------------------------------------------------------------------------  
                                             cutlass-base             streamk                                 
1 threads: ----------------------------------------------------------------------
      MKN=(128x7168x1536)  |                    28.5                  17.1                large K       
      MKN=(128x1536x6144)  |                    10.4                  10.7
      MKN=(128x4096x7168)  |                    19.3                  19.3
      MKN=(128x7168x9216)  |                    32.7                  32.7
      MKN=(128x4608x7168)  |                    20.6                  20.6
      MKN=(128x7168x1024)  |                    26.4                  17                  large K
      MKN=(128x512x7168)   |                    10.4                  10.8
      MKN=(256x7168x4096)  |                    28.9                  29.3
      MKN=(256x7168x2048)  |                    29.8                  22                  large K

Times are in microseconds (us).
```

Hongbosherlock committed:
Add streamK for block-quantized CUTLASS kernels

Signed-off-by: leoneo <1320612015@qq.com>

Hongbosherlock committed:
Add streamK for block-quantized CUTLASS kernels

Signed-off-by: leoneo <1320612015@qq.com>

Hongbosherlock committed:
[Kernel]Add streamK for block-quantized CUTLASS kernels

Signed-off-by: leoneo <1320612015@qq.com>

Hongbosherlock committed:
[Kernel]Add streamK for block-quantized CUTLASS kernels

Signed-off-by: leoneo <1320612015@qq.com>

robertgshaw2-redhat commented:
Cool!

LucasWilkinson commented on csrc/quantization/cutlass_w8a8/c3x/scaled_mm_blockwise_sm90_fp8_dispatch.cuh:
```diff
@@ -154,15 +155,87 @@ void cutlass_gemm_caller_blockwise(torch::Tensor& out, torch::Tensor const& a,
-      cutlass_3x_gemm_fp8_blockwise<OutType, 1, 128, 128>>(out, a, b, a_scales,
-                                                           b_scales);
+  auto k = a_scales.size(1);
+  auto n = b_scales.size(1);
```
nit: use the tensor extents not the scales, `a.size(1)`  `b.size(1)`

LucasWilkinson commented on csrc/quantization/cutlass_w8a8/c3x/scaled_mm_blockwise_sm90_fp8_dispatch.cuh:
```diff
@@ -154,15 +155,87 @@ void cutlass_gemm_caller_blockwise(torch::Tensor& out, torch::Tensor const& a,
+                                           torch::Tensor const& a,
+                                           torch::Tensor const& b,
+                                           torch::Tensor const& a_scales,
+                                           torch::Tensor const& b_scales) {
```
nit: alot of this seems repeated from `cutlass_gemm_caller_blockwise`, can we try to unify these?

LucasWilkinson commented on csrc/quantization/cutlass_w8a8/c3x/cutlass_gemm_caller.cuh:
```diff
@@ -53,6 +53,39 @@ void cutlass_gemm_caller(torch::Device device,
+void cutlass_gemm_caller_streamK(
+    torch::Device device, cute::Shape<int, int, int, int> prob_shape,
+    typename GemmKernel::MainloopArguments mainloop_args,
+    typename GemmKernel::EpilogueArguments epilogue_args) {
```
nit: alot of this is repeated from `cutlass_gemm_caller` I think we could unify these by doing something like:
```
template <typename GemmKernel>
 void cutlass_gemm_caller_streamK(
     torch::Device device, cute::Shape<int, int, int, int> prob_shape,
     typename GemmKernel::MainloopArguments mainloop_args,
     typename GemmKernel::EpilogueArguments epilogue_args,
     typename GemmKernel::TileSchedulerArguments scheduler_args = {}) {
```

Hongbosherlock replied:
thanks,I made some attempts initially, and now it's works for me.

LucasWilkinson commented:
@Hongbosherlock Awesome thanks for the contribution! thats some nice speedups! Left a couple nits.

Hongbosherlock commented:
> @Hongbosherlock Awesome thanks for the contribution! thats some nice speedups! Left a couple nits.

Thank you for the feedback! I’ll take a look at the nits and address them ASAP.

Hongbosherlock committed:
fix some nits

Signed-off-by: leoneo <1320612015@qq.com>

Hongbosherlock commented:
Hi @LucasWilkinson 
Thank you for the feedback! I’ve addressed the nits and verified the performance improvements. 
Could you please check if it’s ready to be merged?

LucasWilkinson commented on csrc/quantization/cutlass_w8a8/c3x/scaled_mm_blockwise_sm90_fp8_dispatch.cuh:
```diff
@@ -22,8 +22,9 @@ namespace vllm {
-          int TileSizeM_ = 128, class ClusterShape = Shape<_1, _2, _1>>
+template <typename SchedulerType, typename OutType, int GroupSizeM_,
+          int GroupSizeN_, int GroupSizeK_, int TileSizeM_ = 128,
+          class ClusterShape = Shape<_1, _2, _1>>
```
Sorry one last thing, can we please render this to:
```
template <typename OutType, int GroupSizeM_,
           int GroupSizeN_, int GroupSizeK_, int TileSizeM_ = 128,
           class ClusterShape = Shape<_1, _2, _1>,
           typename SchedulerType = cutlass::gemm::PersistentScheduler>
```
just so we can leave the interface untouched

LucasWilkinson reviewed (commented):
LGTM, thanks for the hardwork!

MichoChan commented:
is faster than triton moe kernel?
